src/ray/rpc/gcs_server/gcs_rpc_client.h

```diff
@@ -221,69 +211,32 @@ class GcsRpcClient {
             bool handle_payload_status>
   void invoke_async_method(
       PrepareAsyncFunction<Service, Request, Reply> prepare_async_function,
-      GrpcClient<Service> &grpc_client,
+      std::shared_ptr<GrpcClient<Service>> grpc_client,
       const std::string &call_name,
       const Request &request,
       const ClientCallback<Reply> &callback,
       const int64_t timeout_ms) {
-    auto executor = new Executor(
-        [callback](const ray::Status &status) { callback(status, Reply()); });
-    auto operation_callback = [this, request, callback, executor, timeout_ms](
-                                  const ray::Status &status, Reply &&reply) {
-      if (status.ok()) {
-        if constexpr (handle_payload_status) {
-          Status st =
-              (reply.status().code() == (int)StatusCode::OK)
-                  ? Status()
-                  : Status(StatusCode(reply.status().code()), reply.status().message());
-          callback(st, std::move(reply));
-        } else {
-          callback(status, std::move(reply));
-        }
-        delete executor;
-      } else if (!IsGrpcRetryableStatus(status)) {
-        callback(status, std::move(reply));
-        delete executor;
-      } else {
-        /* In case of GCS failure, we queue the request and these requests will be */
-        /* executed once GCS is back. */
-        gcs_is_down_ = true;
-        auto request_bytes = request.ByteSizeLong();
-        if (pending_requests_bytes_ + request_bytes >
-            ::RayConfig::instance().gcs_grpc_max_request_queued_max_bytes()) {
-          RAY_LOG(WARNING) << "Pending queue for failed GCS request has reached the "
-                           << "limit. Blocking the current thread until GCS is back";
-          while (gcs_is_down_ && !shutdown_) {
-            CheckChannelStatus(false);
-            std::this_thread::sleep_for(std::chrono::milliseconds(
-                ::RayConfig::instance()
-                    .gcs_client_check_connection_status_interval_milliseconds()));
-          }
-          if (shutdown_) {
-            callback(Status::Disconnected("GCS client has been disconnected."),
-                     std::move(reply));
-            delete executor;
+    retryable_grpc_client_->template CallMethod<Service, Request, Reply>(
+        prepare_async_function,
+        std::move(grpc_client),
+        call_name,
+        request,
+        [callback](const Status &status, Reply &&reply) {
+          if (status.ok()) {
+            if constexpr (handle_payload_status) {
+              Status st = (reply.status().code() == (int)StatusCode::OK)
```

Comment:
- Don't we have `reply.status().ok()` function? If not we should add one.
- Even if you want to compare status code, isn't `==` operator enough?
- Don't use C-style type casting, it's dangerous, please stick to `static_cast`; checkout effective C++